### PR TITLE
feat: Add freezed for code generation

### PR DESCRIPTION
**Pull request by martinneff**
## Status

**READY**

## Breaking Changes

NO

## Description

With the Intellij plugin states and events could only be generated with or without equatable support. I added support for freezed. Users can select from either non, equatable or freezed when generating cubits or blocs. 

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.bloc.intellij_generator_plugin.action;
+
+/**
+ * @author Martin Neff, approppo GmbH (martin.neff@approppo.ch)
+ * @since 13.09.22
+ */
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java:6]:
```suggestion
```

Is this necessary? I'd prefer not to include such comments in the codebase because they don't add much value and anyone can look at git blame for a history of changes and the contributor.

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+public enum BlocStatePackage {
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java:7]:
```suggestion
public enum BlocTemplateType {
```

File: extensions/intellij/intellij_generator_plugin/src/main/resources/templates/bloc_with_equatable/bloc.dart.template [hunk 3]
```diff
@@ -3,12 +3,12 @@ import 'dart:async';
 import 'package:bloc/bloc.dart';
 import 'package:equatable/equatable.dart';
 
-part '${bloc_snake_case}_event.dart';
-part '${bloc_snake_case}_state.dart';
+part '^{bloc_snake_case}_event.dart';
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/resources/templates/bloc_with_equatable/bloc.dart.template:6]: Why are these changes necessary?

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt [hunk 4]
```diff
@@ -34,7 +39,7 @@ abstract class BlocGenerator(private val name: String,
     abstract fun fileName(): String
 
     fun generate(): String {
-        val substitutor = StrSubstitutor(templateValues)
+        val substitutor = StrSubstitutor(templateValues, "^{", "}", '\\')
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt:42]: why is this substitution needed?

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/CubitGenerator.kt [hunk 5]
```diff
@@ -34,7 +39,7 @@ abstract class CubitGenerator(private val name: String,
     abstract fun fileName(): String
 
     fun generate(): String {
-        val substitutor = StrSubstitutor(templateValues)
+        val substitutor = StrSubstitutor(templateValues, "^{", "}", '\\')
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/CubitGenerator.kt:42]: why is this substitution needed?

File: extensions/intellij/intellij_generator_plugin/src/main/resources/templates/bloc_with_equatable/bloc_event.dart.template [hunk 6]
```diff
@@ -1,5 +1,5 @@
-part of '${bloc_snake_case}_bloc.dart';
+part of '^{bloc_snake_case}_bloc.dart';
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/resources/templates/bloc_with_equatable/bloc_event.dart.template:1]: same here, why are these changes needed?

File: extensions/intellij/intellij_generator_plugin/src/main/resources/templates/bloc_with_equatable/bloc_state.dart.template [hunk 7]
```diff
@@ -1,10 +1,10 @@
-part of '${bloc_snake_case}_bloc.dart';
+part of '^{bloc_snake_case}_bloc.dart';
```
Comment (felangel) [inline extensions/intellij/intellij_generator_plugin/src/main/resources/templates/bloc_with_equatable/bloc_state.dart.template:1]: same here, why are these changes needed?

**Review by felangel — CHANGES_REQUESTED** [review]
Thanks for the contribution! See comments and lmk what you think.

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java (hunk 1 above)
Comment (martinneff) [reply extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java:6]: Ah i forgot to remove the file header in Intellij. I'll remove all those comments

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java (hunk 2 above)
Comment (martinneff) [reply extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/action/BlocStatePackage.java:7]: thanks for the suggestion, ill refactor to the new name

File: extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt (hunk 4 above)
Comment (martinneff) [reply extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt:42]: in the template for the freezed code there's a '$' for the mixin class. The StrSubstitutor class uses '$' as prefix ${ and also as escape char. i wasn't able to make the substitution work for a template like `_$${bloc_camel_case}State` to `_$LoginState`. Therefore i changed the prefix to ^{ and the escape char to something else than '$'.

Comment (felangel) [reply extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt:42]: Hmm is it not possible to escape the freezed $ like `\$`?

Comment (martinneff) [reply extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt:42]: Unfortunately not: the following template:
```
import 'package:freezed_annotation/freezed_annotation.dart';

part '${bloc_snake_case}_event.freezed.dart';

@freezed
class ${bloc_pascal_case}Event with _\$${bloc_pascal_case}Event {
  const factory ${bloc_pascal_case}Event.loadData() = ${bloc_pascal_case}LoadDataEvent;
}
```

with the default substitutor `val substitutor = StrSubstitutor(templateValues)`

results in:

```dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'checkin_event.freezed.dart';

@freezed
class CheckinEvent with _\${bloc_pascal_case}Event {
  const factory CheckinEvent.loadData() = CheckinLoadDataEvent;
}
```

If i add two '\' i get two in the generated code...

Comment (felangel) [reply extensions/intellij/intellij_generator_plugin/src/main/java/com/bloc/intellij_generator_plugin/generator/BlocGenerator.kt:42]: weird...that's super annoying 😛

**Comment by martinneff** [comment]
@felangel I'm not sure about how to proceed. Are there any required changes  you'd expect from me? The currently open requested changes cannot be resolved other than changing all the escape signs in the templates.

**Review by felangel — APPROVED** [review]
